### PR TITLE
Transaction versioning in the RuntimeVersion

### PR DESCRIPTION
**Pull request by gavofyork**
This just adds the `transaction_version` field in the `RuntimeVersion`, which just tracks the compatibility of dispatchables, useful for hardware wallets.

Polkadot companion: https://github.com/paritytech/polkadot/pull/987

### CODE REVIEW

File: primitives/version/src/lib.rs [hunk 1]
```diff
@@ -71,6 +71,17 @@ pub struct RuntimeVersion {
...
+	/// All existing dispatches are fully compatible when this number doesn't change. If this
+	/// number changes, then `spec_version` must change, also.
+	///
+	/// This number must change when an existing dispatchable (module ID, dispatch ID) is changed,
+	/// either through an alteration in its user-level semantics, a parameter added/removed/changed,
+	/// a dispatchable being removed, a module being removed, or a dispatchable/module changing its
+	/// index.
+	///
+	/// It need *not* change when a new module is added or when a dispatchable is added.
+	pub transaction_version: u32,
```
Comment (bkchr) [inline primitives/version/src/lib.rs:83]: This needs to move to the bottom, or we brick everything^^

Comment (bkchr) [reply primitives/version/src/lib.rs:83]: Hmm that will not be enough. We need to change the version of the `Core` runtime api and make sure to put the `changed_in` attribute above the `version` function. 

The old `version` function should have the old `RuntimeVersion` struct as return value.

Comment (gavofyork) [reply primitives/version/src/lib.rs:83]: Moved.

**Comment by xlc** [comment]
This is already a transaction version https://github.com/paritytech/substrate/blob/77fd258408abc0bc6cd0a786c8af004edbefc0a0/primitives/runtime/src/generic/unchecked_extrinsic.rs#L30

And this could fix https://github.com/paritytech/substrate/issues/3925 ?

**Comment by jacogr** [comment]
How does the "1" here relate to the "4" (with the applicable signed extensions) we receive via metadata?

**Comment by gavofyork** [comment]
They are quite different things; the `TRANSACTION_VERSION` constant refers to the version code placed in each transaction and is the version of the transaction's data format, taking into account any extensions it might have. Different versions imply that transactions cannot be decoded in the same way. The `transaction_version` field of the `RuntimeVersion` struct, OTOH, refers to the semantics of the dispatchables. Different versions here imply an specific incompatibility within the dispatchables. Generally this will be the removal of a module or a dispatchable within a module. It could also mean an alteration in the argument list.

The specific semantics are:
- When removing or changing a dispatchable/pallet (including reordering) that was available within the first released runtime of the current runtimes's transaction version, then `transaction_version` of the new release must be bumped.
- When adding a dispatchable/pallet, `transaction_version` need not be bumped.

In both cases, `spec_version` must be bumped. This implies that modules which are introduced into a runtime which is released *after* the first runtime of that `transaction_version` may actually be changed or removed without bumping the `transaction_version`. Bumping the `transaction_version` therefore amounts to "baking in the current set of dispatchables as the API".

**Comment by xlc** [comment]
I see. It should be possible to create some automation to detect if we need to bump transaction version by comparing metadata from wasm? Manually tracking it is error prone.

**Comment by gavofyork** [comment]
Yeah that should be possible to automate; you'd just need to check if there's an alteration/removal to existing pallets, dispatchables, including their docs.

**Comment by gnunicorn** [comment]
seems like tests need updating

**Pull request merged**